wp-includes/class-wp-service-workers.php

```diff
@@ -83,6 +220,170 @@ public function register( $handle, $src, $deps = array(), $scope = self::SCOPE_A
+				$route = array( $route );
+			}
+
+			$this->registered_caching_routes[] = array(
+				'route'    => $route, // @todo route is Not the best name in case of array of routes for precaching.
+				'strategy' => $strategy,
+			);
+		} elseif ( ! is_string( $route ) ) {
+			/* translators: %s is caching strategy */
+			$error = sprintf( __( 'Route for the caching strategy %s must be a string.', 'pwa' ), $strategy );
+			_doing_it_wrong( __METHOD__, esc_html( $error, 'pwa' ), '0.2' );
+		} else {
+
+			// Set default to string literal for route.
+			if ( ! isset( $args['regex'] ) ) {
+				$args['regex'] = false;
+			}
+
+			$this->registered_caching_routes[] = array(
+				'route'    => $route,
+				'strategy' => $strategy,
+				'args'     => $args,
+			);
+		}
+	}
+
+	/**
+	 * Register precaching for routes. Gets the hashes of files' contents and adds as revision for each route.
+	 *
+	 * @param array $routes Array of routes.
+	 * @return string Precaching logic.
+	 */
+	protected function register_precaching_for_routes( $routes ) {
+
+		$routes_list = '[';
+
+		foreach ( $routes as $route ) {
+			$validated_path = $this->get_validated_file_path( $route, false );
+			$content        = '';
+
+			// If it's not a file.
+			if ( ! is_wp_error( $validated_path ) ) {
+				$content = @file_get_contents( $validated_path ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents, WordPress.WP.AlternativeFunctions.file_system_read_file_get_contents
+			} else {
+
+				// Maybe it's an URL.
+				$needs_base_url = ! preg_match( '|^(https?:)?//|', $route );
+				$base_url       = site_url();
+
+				if ( $needs_base_url ) {
+					$route = $base_url . $route;
+				}
+
+				// @todo Remove this and add real revisioning logic instead for URLs.
+				$content = 'offline-post';
+			}
+
+			if ( empty( $content ) ) {
+				continue;
+			}
+
+			$hash = md5( $content );
+			if ( '[' !== $routes_list ) {
+				$routes_list .= ',';
+			}
+			$routes_list .= '
+{
+	url: ' . wp_json_encode( $route ) . ',
+	revision: ' . wp_json_encode( $hash ) . ',
+}';
+		}
+
+		if ( '[' === $routes_list ) {
+			return '';
+		}
+
+		$routes_list .= ']';
+
+		return 'wp.serviceWorker.precaching.precacheAndRoute(' . $routes_list . ");\n";
+	}
+
+	/**
+	 * Register caching strategy for route.
+	 *
+	 * @param string $route Route.
+	 * @param int    $strategy Caching strategy.
+	 * @param array  $args Array of args, can be cache_name, max_age, max_entries.
+	 * @return string Script.
+	 */
+	protected function register_caching_strategy_for_route( $route, $strategy, $args ) {
+		$script = '';
+
+		if ( isset( $args['cache_name'] ) ) {
+			$script .= 'const args = {
+	cacheName: ' . wp_json_encode( $args['cache_name'] ) . '
+};';
+		}
+		if ( isset( $args['max_entries'] ) || isset( $args['max_age'] ) ) {
+			$script .= 'args.plugins = [
+	new wp.serviceWorker.expiration.Plugin({';
+			if ( isset( $args['max_age'] ) ) {
```

Comment:
The the `max_age` shouldn't be hard-coded here in `$args` I don't think. Instead, `$args` should have a `plugins` array like:

```php
array(
    'plugins' => array(
        'expiration' => array(
            'maxAgeSeconds' => 123,
            'maxEntries' => 45,
        ),
    )
)
```

This would then allow _any_ plugin to be referenced and passing all of the args that the plugins accept. The `$args` would be whatever the caching strategy `options` are accepted, for example: https://developers.google.com/web/tools/workbox/reference-docs/latest/workbox.strategies.CacheFirst#constructor_1

So `$args` should be renamed `$strategy_args` for clarity.

Reply:
That being the case, I think the `$regex` should not be passed in the `$args` now. Perhaps there should be a separate methods like `register_cached_route_pattern()` and `register_cached_route_url()` to differentiate between the two types of args allowed?

Reply:
Made the general Workbox related changes within 6c2592f in #43.

Makes sense that the plugins shouldn't be hardcoded and it should be possible to add other plugins, too. Made these changes within 6c2592f.

On regex: what do you think of just adding an additional param to the method so that if a developer would switch from between regex/url then only a param would require changing instead of the method? For example: `register_cached_route( $route, $strategy, $strategy_args = array(), $is_regex = false )`.
Thinking that since the only difference it makes when registering routes to `self::$registered_caching_routes` would be a change if `is_regex` is `true/false` for the relevant route, then maybe an additional param would be enough.

Or are you thinking that maybe it's more straightforward and better structured to have two separate methods? Thoughts?

(Added `$is_regex` param as a temporary change within 6c2592f to move this out of the `$strategy_args`, will it change it accordingly.)

Reply:
Makes sense. Split the method into two within 0fd9320.